### PR TITLE
stub: Have ClientCalls.ThreadlessExecutor reject Runnables after end of RPC

### LINKED ISSUE #3557: ClientCalls.ThreadlessExecutor should reject Runnables after end of RPC
As seen in #3537, the executor was being used for multiple RPCs which caused some Runnables to be queued but never run. We are in a position to know they will never execute though, so it'd be nice to have a (possibly unreliable) detection of rejected Runnables to notice issues like this earlier.

This could be accomplished with just a volatile `shutdown` boolean that is checked before `execute()` and a `shutdown()` method. That's racy, but we're just wanting it to detect bugs, not provide correctness.

### PR DESCRIPTION
**Pull request by njhill**
Changes originally proposed as part of #7106.

Fixes #3557

### CODE REVIEW

File: stub/src/main/java/io/grpc/stub/ClientCalls.java [hunk 1]
```diff
@@ -712,6 +717,8 @@ void onStart() {
       implements Executor {
     private static final Logger log = Logger.getLogger(ThreadlessExecutor.class.getName());
 
+    private static final Thread SHUTDOWN = new Thread(); // sentinel
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/ClientCalls.java:720]: Creating a Thread worries me and seems like it is asking for trouble. Let's just make `waiter` an Object, which will either be null, a Thread, or a sentinel.

File: stub/src/main/java/io/grpc/stub/ClientCalls.java [hunk 2]
```diff
@@ -626,6 +628,9 @@ private Object waitForNext() {
               // Now wait for onClose() to be called, so interceptors can clean up
             }
           }
+          if (next == this) {
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/ClientCalls.java:631]: `|| next instanceof StatusRuntimeException`

**Comment by ejona86** [comment]
@voidzcy, you self-assigned to review the earlier PR that this one is a subset of. Do you want to review this one as well?

**Review by ejona86 — APPROVED** [review]

**Review by voidzcy — APPROVED** [review]
Looks nice.

**Comment by ejona86** [comment]
@njhill, thank you!

**Pull request merged**